feat(python): support for 3.12

vkurilin opened the pull request:
Closes #157

francium commented:
Thanks.

I think there's a few files however we still need to update.
See this PR for 3.11, https://github.com/rustedpy/result/pull/107/files, for reference

vkurilin commented:
Thanks!

I've added `[Unreleased]` record to the changelog and fixed `tox.ini`, but I don’t understand how to correctly use the `test_result.yaml` file

francium commented:
You should be able to run `pytest` and it should pick up the yaml test. But as long as the PR passes checks, I don't know if we need any changes to the yaml file.

Thanks, I'll review this and test this myself when I get the chance.

vkurilin commented:
Oh, cool, I didn't even know there was a way to run tests like that

```
type-checking/test_result.yml::failure_lash PASSED
type-checking/test_result.yml::covariance PASSED
type-checking/test_result.yml::map_ok_err PASSED
type-checking/test_result.yml::map_result PASSED
type-checking/test_result.yml::map_result PASSED
```

These tests pass, they show that we have fixed everything that was needed =)

francium commented:
Thanks for the help @vkurilin
